packages/terra-application-utility/src/header/_HeaderUtilityMenu.jsx

```diff
@@ -0,0 +1,256 @@
+      currentKey: props.initialSelectedKey,
+      previousKeyStack: [],
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState({
+      map: processMenuItems(nextProps.menuItems),
+    });
+  }
+
+  getItem(key) {
+    return this.state.map.get(key);
+  }
+
+  buildItem(key) {
+    const item = this.getItem(key);
+    const chevron = hasChevron(item);
+    return (
+      <MenuItem
+        key={key}
+        itemKey={key}
+        id={item.id}
+        title={item.title}
+        content={item.content}
+        contentLocation={item.contentLocation}
+        isSelected={item.isSelected}
+        isSelectable={item.isSelectable}
+        hasChevron={chevron}
+        onChange={this.handleOnChange}
+      />
+    );
+  }
+
+  buildListContent(currentItem) {
+    if (currentItem && currentItem.childKeys && currentItem.childKeys.length) {
+      return (
+        <ul className={cx('body')}>
+          {currentItem.childKeys.map((key) => {
+            let item = null;
+            if (this.getItem(key).contentLocation === Utils.LOCATIONS.BODY) {
+              item = this.buildItem(key);
+            }
+            return item;
+          })}
+        </ul>
+      );
+    }
+    return null;
+  }
+
+  buildFooterContent(currentItem) {
+    if (currentItem && currentItem.childKeys && currentItem.childKeys.length) {
+      return currentItem.childKeys.map((key) => {
+        let item = null;
+        if (this.getItem(key).contentLocation === Utils.LOCATIONS.FOOTER) {
+          item = this.buildItem(key);
+        }
+        return item;
+      });
+    }
+    return null;
+  }
+
+  /**
+   * Function to trigger when an item is selected.
+   * 1. Has children: navigate to the next page
+   * 2. Toggles: trigger onChange without closing the menu.
+   * 3. Endpoint: close menu and trigger onChange.
+   * @param {*} event
+   * @param {*} key
+   */
+  handleOnChange(event, key) {
+    const childKeys = this.getItem(key).childKeys;
+    const item = this.getItem(key);
+    if (typeof childKeys !== 'undefined' && childKeys.length > 0) {
+      this.toggleIsSelected(key);
+      this.setState({
+        previousKey: this.push(this.state.currentKey),
+        currentKey: key,
+      });
+    } else {
+      this.toggleIsSelected(key);
+      if (item.isSelectable !== true) {
+        this.props.onRequestClose();
+      }
+      this.props.onChange(event, key);
+    }
+  }
+
+  handleRequestBack() {
+    this.setState({ currentKey: this.pop() });
+  }
+
+  pop() {
+    const newStack = this.state.previousKeyStack.slice();
+    const lastKey = newStack.pop();
+    this.setState({ previousKeyStack: newStack });
+    return lastKey;
+  }
+
+  push(key) {
+    const newStack = this.state.previousKeyStack.slice();
+    newStack.push(key);
+    this.setState({ previousKeyStack: newStack });
+  }
+
+  toggleIsSelected(key) {
+    const item = this.getItem(key);
+    item.isSelected = !item.isSelected;
+    this.state.map.set(key, item);
```

Comment:
We should not be manipulating the config, this should be handled by the consumer updated the selected state of the passed config.